Killer damage HUD (replacing menu)

nullsystem opened the pull request:
<!--
Before submitting a pull request, ensure the following has been done:
* The branch has been tested with the latest master changes rebased in
* Fill in the descriptions, link the issues, and put in tags appropriate to the PR
* Update any documentation and comments if needed
* For WIP/Work in Progress PRs, use the Draft PR feature
-->

## Description
<!--
Put in description here...
-->
* Replace the menu-based killer/damage info with HUD
* Right-side, generally gives slight more killer/damage info than menu
* Keybind toggle
* Pagination
* Convar+settings on how it pops up/interacts (`cl_neo_kdinfo_toggletype`)

## Toolchain
<!--
If this is documentation only update, just remove the whole Toolchain section
NOTE: It's not needed for all to be filled in, just keep the toolchain/OS lines this PR been worked on
- Windows MSVC VS2022
-->
- Linux GCC Distro Native Gentoo/GCC 14.3.0

## Linked Issues
<!--
Applying issues here will auto-link the PR to its related issues if starting with "resolves".
If there's a related PR but don't want to resolve/close the issue, mark them with "related".

See: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->
* fixes #667

sunzenshen commented on src/game/client/neo/c_neo_player.cpp:
```diff
@@ -142,18 +145,52 @@ static void __MsgFunc_DamageInfo(bf_read& msg)
+			g_pVGuiLocalize->ConvertANSIToUnicode(neoAttacker->GetNeoPlayerName(),
+												  g_neoKDmgInfos.wszKillerName,
+												  sizeof(g_neoKDmgInfos.wszKillerName));
+			g_neoKDmgInfos.iKillerNameSize = wcslen(g_neoKDmgInfos.wszKillerName);
```
Maybe we want to standardize on using V_wcslen?

nullsystem replied:
`GetNeoPlayerName` and `wszKillerName` both have `MAX_PLAYER_NAME_LENGTH + 1` as array length and `g_neoKDmgInfos` zeroed out before this section, so don't think I'll be too concerned about it.

Rainyan replied:
> Just a sanity check question, but is there any chance that ConvernANSIToUnicode could error out and not write a null terminator, causing this len check to read past the buffer?

For Linux, the call resolves to: https://github.com/NeotokyoRebuild/neo/blob/f5cf0e8c24ca28c5c7b8ac0670840ffafdceb538/src/tier1/strtools_unicode.cpp#L233 which looks like it'll null-terminate the output even in error state.

For Windows, it resolves to: https://learn.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-multibytetowidechar which has the following trap:
> MultiByteToWideChar does not null-terminate an output string if the input string length is explicitly specified without a terminating null character.

But for reasons nullsystem stated above, probably not an issue here either way.

sunzenshen commented on src/game/client/neo/ui/neo_hud_killer_damage_info.cpp:
```diff
@@ -0,0 +1,346 @@
+			m_uiCtx.eLabelTextStyle = NeoUI::TEXTSTYLE_CENTER;
+
+			const int iCurPageStart = m_iCurPage * PAGE_MAX_SHOW;
+			const int iCurPageEnd = min(m_iAttackersTotalsSize, iCurPageStart + PAGE_MAX_SHOW);
```
Maybe use Min instead

sunzenshen commented on src/game/client/neo/ui/neo_hud_killer_damage_info.cpp:
```diff
@@ -0,0 +1,346 @@
+
+	m_uiCtx.layout.iDefRowTall = tall / 27;
+	m_uiCtx.iMarginX = wide / 192;
+	m_uiCtx.iMarginY = tall / 108;
```
I have no idea where these numbers come from, so maybe these could be constants?

nullsystem replied:
TODO for issue #1253

nullsystem commented:
Rebased for master, but also altered some stuff slightly:

* Made sure to only update HUD infos when necessary
* Changed default from match session to always
* Add bracket to toggle label on round vs match type toggle
* Minor refactor/cleanups

Rainyan reviewed (commented):
![Screenshot](https://github.com/user-attachments/assets/4253be60-778b-4803-83f0-b88403572741)

Would it be possible for us to completely cut/not render the killfeed for the pixels overlapping the KID panel? Or if that's unreasonably much effort (not too familiar with the UI details), perhaps having the KID panel bg take full (or fuller) alpha opacity, and rendering the killfeed under it, so that it doesn't distract the reading as much at areas where the two overlap.

Another possible option would be to make it such that:

* If the Killer Info panel is open
* And if there's a killfeed item that would overlap with it
* Expire any upper killfeed items for as long as it takes to resolve the overlap

in which case we could dodge this problem. But I'm not sure which is the best approach.

And I guess another option would be to simply move the panel way more to the left, such that it will not overlap no matter what carnage is happening in the killfeed. Maybe this would be cleaner actually?

---

It's also a bit complicated to understand since it's quite busy visually, and many elements/info is packed very close together, but that could maybe be a thing to improve in the future for some visual design/UX pass, instead of getting too fixated on it for this PR.

---

One thing I might like to see changed is the `Killed by: <weapon>` language, since we have two of them appearing at the same time:
* `Killed by <Player name>`
* `Killed by: <Weapon name>`

which makes it a bit hard to read, since they're so similar.

Maybe it'd be better to have:

* `Killed by <Player name>`
* `Weapon: <weapon name>`

AdamTadeusz commented:
> ![Screenshot](https://private-user-images.githubusercontent.com/6595066/478151505-4253be60-778b-4803-83f0-b88403572741.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUyMDM3ODcsIm5iZiI6MTc1NTIwMzQ4NywicGF0aCI6Ii82NTk1MDY2LzQ3ODE1MTUwNS00MjUzYmU2MC03NzhiLTQ4MDMtODNmMC1iODg0MDM1NzI3NDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MTRUMjAzMTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjM1MGFlNTZhNjU1NzIxMGY4ZGM3YjE1ZDE4OTA0ZmEzMmEyOTRkYzRlYWY1M2FhNTc1MjA3NWIwOTZkYzAwZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.OtuXeeB6xeNjaAo2folu5BDmVr3JZvvpZ3-jfgz7GQY)
> 
> Would it be possible for us to completely cut/not render the killfeed for the pixels overlapping the KID panel? Or if that's unreasonably much effort (not too familiar with the UI details), perhaps having the KID panel bg take full (or fuller) alpha opacity, and rendering the killfeed under it, so that it doesn't distract the reading as much at areas where the two overlap.
> 
> Another possible option would be to make it such that:
> 
>     * If the Killer Info panel is open
> 
>     * And if there's a killfeed item that would overlap with it
> 
>     * Expire any upper killfeed items for as long as it takes to resolve the overlap
> 
> 
> in which case we could dodge this problem. But I'm not sure which is the best approach.
> 
> And I guess another option would be to simply move the panel way more to the left, such that it will not overlap no matter what carnage is happening in the killfeed. Maybe this would be cleaner actually?
> 
> It's also a bit complicated to understand since it's quite busy visually, and many elements/info is packed very close together, but that could maybe be a thing to improve in the future for some visual design/UX pass, instead of getting too fixated on it for this PR.
> 
> One thing I might like to see changed is the `Killed by: <weapon>` language, since we have two of them appearing at the same time:
> 
>     * `Killed by <Player name>`
> 
>     * `Killed by: <Weapon name>`
> 
> 
> which makes it a bit hard to read, since they're so similar.
> 
> Maybe it'd be better to have:
> 
>     * `Killed by <Player name>`
> 
>     * `Weapon: <weapon name>`

A few extra pixels of space between each player in the report would be nice

sunzenshen commented on src/game/server/neo/neo_player.cpp:
```diff
@@ -1804,182 +1768,31 @@ bool CNEO_Player::BecomeRagdollOnClient( const Vector &force )
-			killedWithName = killedWithInflictor ? (inflictorIsPlayer ? neoAttacker->m_hActiveWeapon->GetPrintName() : killedWithInflictor->GetDebugName()) : "";
-			if (!Q_strcmp(killedWithName, "neo_grenade_frag")) { killedWithName = "Frag Grenade"; }
-			if (!Q_strcmp(killedWithName, "neo_deployed_detpack")) { killedWithName = "Remote Detpack"; }
+			if (neoAttacker->entindex() != entindex())
```
This if logic might be redundant if line 1782 already checks `neoAttacker->entindex()`

Did you mean to check the assignment of attackerIdx somehow?

AdamTadeusz commented on src/game/client/neo/ui/neo_hud_killer_damage_info.cpp:
```diff
@@ -0,0 +1,559 @@
+	SetVisible(false);
+
+	SetFgColor(COLOR_TRANSPARENT);
+	SetBgColor(COLOR_TRANSPARENT);
```
I think we've done this in a few hud elements but its actually unnecessary. In cdll_client_int.cpp once gHud.Init(); initialises all the panels, IGameSystem::InitAllSystems() a few lines down IGameSystem::InitAllSystems reloads the scheme of the base viewport which calls ApplySchemeSettings on all of its children. Then again maybe its safer to do this, I guess not all elements have to be a child of the baseviewport, and that could change for this element somewhere down the line at which point this would break https://developer.valvesoftware.com/wiki/VGUI_Documentation

nullsystem replied:
I think we did this way in the past because the HUD elements were giving a darker tint to the entire screen? Either way just file this as something to take a look if doing all this really needed or not, but not in this PR.

AdamTadeusz commented on src/game/client/neo/ui/neo_hud_killer_damage_info.cpp:
```diff
@@ -0,0 +1,559 @@
+		m_bShowInfo = false;
+		pwszToggleTypeLabel = L"round";
+		break;
+	case KDMGINFO_TOGGLETYPE_MATCH:
```
rain suggested a [[fallthrough]] once elsewhere, should that be used here?

AdamTadeusz commented on src/game/client/neo/ui/neo_hud_killer_damage_info.cpp:
```diff
@@ -0,0 +1,559 @@
+			const char *dmgerClass = GetNeoClassName(neoAttacker->GetClass());
+
+			char infoLine[128];
+			if (totals.dealtDmgs > 0 && totals.takenDmgs > 0)
```
<img width="1004" height="1130" alt="image" src="https://github.com/user-attachments/assets/b2e863f8-0e6e-46cc-b985-114970b9191a" />

Looks like this nicely skips the info in the killedByLine if we died to this player, but in other instances it looks like it indeed might be skipping the first person, unless the reason for that is that the console skips the last bullet to have killed the local player or the last bullet to have been fired before the round ends?

AdamTadeusz replied:
Also whats with those question marks

nullsystem replied:
Good catch, it's supposed to be attackerInfo not total. Not sure how it went like that probably just some too quick edits.

AdamTadeusz replied:
hmm actually yeah in the second case its also wrong, I guess the intention was to have the list after the question marks be exhaustive then

nullsystem replied:
The question marks seems like corrupt data, but it doesn't seem to make much sense why it happens for you?

nullsystem replied:
Question marks also related to the usage of totals issue.

AdamTadeusz reviewed (changes requested):
Seems fine mostly. Some design suggestions that you can ignore, but I think it would be good if the mismatch between the information in the panel and in the console could be fixed, also confused about the indexing like @sunzenshen mentioned around line 529. It seems like its handled separately before the for loop but like I showed on that image in the third test it seems that a person is missing, but that could be linked to the last bullet mismatch between the panel and console

nullsystem commented on game/neo/scripts/kb_def.lst:
```diff
@@ -44,6 +44,9 @@
 "F10" "joinstar 0"
+"F11" "kdinfo_toggle"
+"p" "kdinfo_page_prev"
+"n" "kdinfo_page_next"
```
If the arrow keys used in the main menu this is going to collide with the already existing slider and left-right widgets movement, so it should not. I'd also rather not mixup keybindings with main menu bindings, they're both should be separated and will screw up TextEdit if someone binds those to alphanumerical keys.

sunzenshen reviewed (approved):
LGTM.  Damage from Jeff isn't counted (blank results if you only took damage from non-player entities), but I think that kind of detail is optional.  Otherwise results look good.

nullsystem commented:
@AdamTadeusz @sunzenshen Will look at it separately: #1275
